### DISCUSSION TITLE
Comparison with cargo-semver-checks

### DISCUSSION THREAD

**Discussion by AnthonyMikh** [cargo-public-api/cargo-public-api#629] [Q&A] [answered]
One of the goals of this project is

> ...Detect breaking API changes and semver violations via CI or a CLI

However detecting semver violations in Rust is [tricky](https://predr.ag/blog/some-rust-breaking-changes-do-not-require-major-version) since it involves a ton of edge cases. This knowledge is already (mostly) incorporated into [cargo-semver-checks](https://github.com/obi1kenobi/cargo-semver-checks). Since apparently this project and that do not share any kind of semver-violating code detection logic but use the same source of truth (rustdoc-generated JSON), for me personally it looks like some kind of duplicated effort.

What is advantage of `cargo-public-api` over `cargo-semver-checks`? Does it make sense do consolidate the effort in order to avoid independently rediscovering intricacies of Rust semver checks?

**Comment by Enselic** [comment]
I think one advantage of this project is that it detects "all" kinds of semver regressions "without extra effort". Whereas for cargo-semver-checks, one has to write a detector for each kind of semver breakage. If no one has written a detector for a certain kind of semver breakage, it will go undetected. This problem does not exist for cargo-public-api, because it detects "everything".

That said, I think there might be a few cases that cargo-server-checks detects that the cargo-public-api suite of crates do not detect. But I don't know of any particular examples.

Sorry to ping you directly @obi1kenobi, but maybe you can correct me if I'm wrong?

**Reply by obi1kenobi** [reply]
There are certainly many cases `cargo-semver-checks` doesn't catch yet. The easiest one is this:
```rust
pub fn example(x: i64) {}
// changes to
pub fn example(x: String) {}
```
None of our checks have the ability to read function argument or return types, so this would sneak by. That said, we've ~doubled the number of checks since the start of this year — we have 90+ right now, and more get added every week.

## Tricky cases `cargo-semver-checks` catches

Here are a few tricky edge cases that `cargo-semver-checks` catches, and a bonus scary case we don't catch (yet!). The challenge in each case is both spotting the change and also explaining why it might or might not be breaking. Sometimes, the item being added or changed is *not* the item that suffers a breaking change! Since many of these aren't obvious, I'm including a runnable doctest that proves each change is breaking in the manner stated by each example.

I haven't looked at `cargo-public-api` closely enough to know how / if they would be flagged. If they are indeed flagged, I'd be curious if the `cargo-public-api` report includes sufficient context to know whether the change is SemVer-major or not.

### Re-export edge cases

```rust
/// ` ``rust
/// let _ = playground::Example;
/// ` ``
#[doc(hidden)]
mod internals {
    mod inner {
        pub struct Example;
    }

    pub use inner::*;
}

// Re-export everything at top level.
// No item in the resulting
// `playground::Example` path is `doc(hidden)`,
// so `Example` is public API.
//
// Removing either this glob-re-export,
// or `Example` itself, would be a breaking change.
pub use internals::*;
```
[playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=c203ed5c4b3e208ef2ad9de84a771d42)

### Losing `dyn Trait` ability

```rust
/// ` ``rust
/// fn example(value: &dyn playground::Trait) {}
/// ` ``
pub trait Trait {
    // Uncommenting the next line is a major breaking change,
    // but only if the trait *previously* was dyn-safe:
    // https://doc.rust-lang.org/reference/items/traits.html#object-safety
    //
    // const ZERO: i64 = 0;
}
```
[playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=710c31a476bb3e4fad2f4ff9f0c0ec54)

### Change of discriminant value in an enum variant

```rust
/// ` ``rust
/// assert_eq!(
///     playground::Example::Third as isize, 2
/// )
/// ` ``
#[non_exhaustive]
pub enum Example {
    First,
    
    // Replacing the below line with
    // the commented-out one below it
    // is a major breaking change on both `Second` and *`Third`*,
    // since it changes *both* discriminants.
    Second,
    // Second = 5,
    
    // Adding the line below is also a major breaking change
    // on *Third*, even though this enum is `#[non_exhaustive]`.
    // The breakage is that `Third`'s discriminant changes.
    //
    // Extra,
    
    Third,
}

// Also note that all enum discriminant changes
// affect the ordering generated by `#[derive(PartialOrd, Ord)]`,
// so all `.sort()` calls of types containing such an enum
// would suffer a breaking change in behavior!
```
[playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=45ad1d20254ac1e0c4fd5dd8a582794d)

### Bonus 1: Changes that are only breaking if a trait is sealed

Certain traits are only breaking if a trait is not "sealed", where a sealed trait is one where a downstream crate cannot have an `impl` of that trait. Determining if a trait is sealed is *very complex* but `cargo-semver-checks` implements the full algorithm and can correctly triage complex cases like this:
```diff,rust
 pub trait Example: Super {
     fn frobnicate(x: crate::defs::Value);

+    // Enables new functionality in our app.
+    fn tweak();
 }
```

Here, the change _might or might not_ be breaking — it depends on the specifics of the `Super` trait, and the `crate::defs::Value` type. More details on that here: https://predr.ag/blog/is-this-trait-sealed-or-not-sealed/

### Bonus 2: Nasty glob re-export shadowing

`cargo-semver-checks` doesn't catch this _yet_, but we have 99% of the necessary infra to catch it in the future. It might be quite hard to spot this from an API report though.

```rust
/// ` ``rust
/// let _: playground::example = playground::example();
/// ` ``
mod internals {
    mod inner {
        #[allow(non_camel_case_types)]
        pub struct example();
    }

    pub use inner::*;
    
    // Adding this function is a major breaking change.
    // If you uncomment it, the doctest will break.
    //
    // pub fn example() {}
}

pub use internals::*;
```
[playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e18cc12e26d5477327632a417f6f58c7)

Uncommenting the function gives it priority over the glob-re-exported items. But the glob-re-exported items included both the `example` type and its `example` constructor function, and the constructor gets shadowed by the explicitly-defined function. That shadowing is a major breaking change.

**Reply by Enselic** [reply]
Very interesting examples, thanks 👍 For reference, here is how `cargo public-api` displays the changes:

### Initial example

This is what `cargo public-api diff` shows:

```
Changed items in the public API
===============================
-pub fn diff_test::example(x: i64)
+pub fn diff_test::example(x: alloc::string::String)
```

which is pretty clear semver breakage IMHO.

### Re-export edge cases

Here we have a `cargo public-api` bug. With the initial code, `Example` is not listed as part of the public API, but `cargo doc` shows it as part of the public API. If we fix that, I am confident that `cargo public-api diff` will show that `Example` was removed with `cargo public-api diff`.


### Losing `dyn Trait` ability

This is what `cargo public-api diff` shows:

```
Added items to the public API
=============================
+pub const diff_test::Trait::ZERO: i64
```

and personally I would not think this was a semver breaking change, so in this case `cargo-semver-checks` is very helpful.


### Change of discriminant value in an enum variant

This is what `cargo public-api diff` shows:

```
Changed items in the public API
===============================
-pub diff_test::Example::Second
+pub diff_test::Example::Second = 5
```

which I think clearly indicates "think twice about semver breakage here".


### Bonus 1: Changes that are only breaking if a trait is sealed

The code is not self-contained so I am too lazy to try (sorry!).


### Bonus 2: Nasty glob re-export shadowing

This is what `cargo public-api diff` shows:

```
Added items to the public API
=============================
+pub fn diff_test::example()
```

and this case is indeed very nasty, because previously I thought this diff never was semver breaking, but I agree that in your example it is. Here `cargo-semver-checks` is also very helpful.

**Reply by obi1kenobi** [reply]
That all makes sense, thank you for trying all the examples and posting the results!

Our [sealed traits test cases file](https://github.com/obi1kenobi/trustfall-rustdoc-adapter/blob/rustdoc-v34/test_crates/sealed_traits/src/lib.rs) has a lot of cursed examples of "I can't believe that trait isn't sealed" situations.

Here are some self-contained examples inspired by it.

### Not breaking: `Example` is sealed because `Super` is private
```rust
pub mod defs {
    pub struct Value;
}

trait Super {}

pub trait Example: Super {
    fn frobnicate(x: crate::defs::Value);

    // This method gets added.
    fn tweak();
}
```

### Not breaking: `Example` is sealed because `Value` cannot be imported
```rust
// This module is private now.
mod defs {
    pub struct Value;
}

pub trait Super {}

pub trait Example: Super {
    fn frobnicate(x: crate::defs::Value);

    // This method gets added.
    fn tweak();
}
```

### Not breaking: `Example` is sealed because `Sealed` is not importable
```rust
pub mod defs {
    pub struct Value;
}

mod private {
    pub trait Sealed {}
}

pub trait Super : private::Sealed {}

pub trait Example: Super {
    fn frobnicate(x: crate::defs::Value);

    // This method gets added.
    fn tweak();
}
```

### Breaking: `Example` is not sealed because `NotSealed` has a blanket impl
```rust
pub mod defs {
    pub struct Value;
}

mod private {
    pub trait NotSealed {}

    // This blanket impl ruins what 
    // would otherwise be a sealed trait.
    impl<T> NotSealed for T {}
}

pub trait Super: private::NotSealed {}

pub trait Example: Super {
    fn frobnicate(x: crate::defs::Value);

    // This method gets added.
    fn tweak();
}
```

### Not breaking: The supertrait's blanket impl has a bound on a sealed trait
```rust
pub mod defs {
    pub struct Value;
}

mod private {
    pub trait Sealed {}

    pub trait WithBlanket {}

    // Downstream crates can't use this blanket impl
    // because they can't match the bound: it's on a sealed trait.
    impl<T: Sealed> WithBlanket for T {}
}

pub trait Super: private::WithBlanket {}

pub trait Example: Super {
    fn frobnicate(x: crate::defs::Value);

    // This method gets added.
    fn tweak();
}
```

### Breaking: The supertrait's blanket impl has a bound on an implementable trait
```rust
pub mod defs {
    pub struct Value;
}

mod private {
    pub trait WithBlanket {}

    // Downstream crates *can* use this blanket impl,
    // because they just need to `impl Clone` first.
    impl<T: Clone> WithBlanket for T {}
}

pub trait Super: private::WithBlanket {}

pub trait Example: Super {
    fn frobnicate(x: crate::defs::Value);

    // This method gets added.
    fn tweak();
}
```

**Comment by obi1kenobi** [comment, marked as answer]
There are two important properties to consider in a breakage-detection system:
- **completeness**: "if a breaking change existed, it was found and reported" i.e. "no false negatives"
- **soundness**: "all findings reported as breaking are indeed breaking changes" i.e. "no false positives"

As with most nice things in computer science, there's a law called [Rice's theorem](https://en.wikipedia.org/wiki/Rice%27s_theorem) that says we can't have both. We have to pick at most one.

`cargo-semver-checks` broadly speaking trades off completeness for soundness. False positives are always bugs, and false negatives are inevitable.

`cargo-public-api` broadly speaking makes the opposite choice. Some false positives are inevitable, and false negatives are much more rare.

The right tool for the job is dependent on how much work you want to put in, and whether you are more worried about having false-positives or false-negatives:
- For general-purpose use by non-SemVer-experts, false-positives are usually much more harmful than false-negatives. They might incorrectly block CI and frustrate users, and might "teach" users the wrong thing. `cargo-semver-checks` presents a more automated process where it doesn't expect the user to be a SemVer expert.
- In contrast, there are workloads where it's preferable to dedicate a skilled SemVer reviewer to closely review all API changes than to risk anything slipping by. With some work, such an expert can usually distinguish false-positives from real breakage (though the work [might sometimes be substantial](https://predr.ag/blog/is-this-trait-sealed-or-not-sealed/)) and in that case it's better to give them a full log of the changes than to risk overlooking something. This is a great use case for `cargo-public-api`.

Both are valid use cases. Neither tool can really replace the other. And I feel @Enselic and I have a positive relationship where our shared experiences have made both our tools better, so it doesn't feel like a duplication of effort to me :+1:

**Reply by Enselic** [reply]
This is a phenomenal way to summarize the difference and I recommend this to be marked as the answer. Thanks Predrag!

And I agree our tools complement each other in a nice way, and I hope we can continue our positive relationship ❤️
